dNR: include the rest of the parameters on the MatchedRuleInfoDebug object

elijahsawyers opened the pull request:
#### b222c16121ee83743e34c9fb60f221bff3a63639
<pre>
dNR: include the rest of the parameters on the MatchedRuleInfoDebug object
<a href="https://bugs.webkit.org/show_bug.cgi?id=297881">https://bugs.webkit.org/show_bug.cgi?id=297881</a>
<a href="https://rdar.apple.com/157880177">rdar://157880177</a>

Reviewed by Brian Weinstein.

This patch makes the necessary changes to include the rest of the information on the MatchedRuleInfo
object that&apos;s passed to the onRuleMatchedDebug listeners.

Most of the request properties could easily be set with the information that is already available in
ContentExtensionsBackend::processContentRuleListsForLoad. However, we are missing the rulesetId
property of the rule object, and it&apos;s not readily available because compiled content rule lists do
not track which ruleset a rule belongs to. This patch updates the ReportIdentifierAction to inherit
from ActionWithStringMetadata so that we can serialize its ruleset identifier too in addition to the
rule identifier.

For cleaner code, we&apos;ve chosen to serialize the ruleset identifier repeatedly because most popular
dNR web extensions appear to choose fairly short strings for the IDs. If this proves to consume too
much memory in the future, we can update the strings to be serialized once and store an index to it
in the ReportIdentifierAction.

The identifiers of static rulesets are set in the manifest. Session and dynamic rules have the
identifier, &quot;_session&quot; and &quot;_dynamic,&quot; respectively.

For information about the properties that we&apos;ve added, take a look at:
<a href="https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/onRuleMatchedDebug#listener">https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/onRuleMatchedDebug#listener</a>

It&apos;s worth noting that we aren&apos;t including the parentDocumentId right now because there is some work
to be done to support it with site isolation turned on. That work is tracked with: <a href="https://rdar.apple.com/159289161">rdar://159289161</a>

Added a new tests and updated existing tests to validate this fix.

* Source/WebCore/contentextensions/ContentExtensionActions.h:
(WebCore::ContentExtensions::ReportIdentifierAction::ReportIdentifierAction):
(WebCore::ContentExtensions::ReportIdentifierAction::isolatedCopy const):
(WebCore::ContentExtensions::ReportIdentifierAction::isolatedCopy):
(WebCore::ContentExtensions::ReportIdentifierAction::serialize const):
(WebCore::ContentExtensions::ReportIdentifierAction::deserialize):
(WebCore::ContentExtensions::ReportIdentifierAction::serializedLength):
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::loadRuleIdentifier):
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
* Source/WebCore/contentextensions/ContentRuleListMatchedRule.h:
(WebCore::ContentRuleListMatchedRule::ContentRuleListMatchedRule):
* Source/WebCore/loader/ResourceLoadInfo.cpp:
(WebCore::ContentExtensions::resourceTypeToStringForMatchedRule):
* Source/WebCore/loader/ResourceLoadInfo.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules):
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
(WebKit::WebExtensionContext::handleContentRuleListMatchedRule):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::handleContentRuleListMatchedRule):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:condition:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm:
(WebKit::for):
(+[_WKWebExtensionDeclarativeNetRequestTranslator jsonObjectsFromData:errorStrings:]):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIDeclarativeNetRequest::updateDynamicRules):
(WebKit::WebExtensionAPIDeclarativeNetRequest::updateSessionRules):
* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::TEST_F(ContentExtensionTest, ReportIdentifierAction)):

Canonical link: <a href="https://commits.webkit.org/299333@main">https://commits.webkit.org/299333@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b8727def53f3eb1eca74231ba75e72a6cfa838c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118647 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38328 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28979 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124827 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70707 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/58343840-0502-4fc1-bd40-c094b5b3baa3) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120525 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39024 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46910 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/90046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7d8d5821-01da-42db-a282-4cef325af8ac) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121600 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/31088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/106376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70551 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/26d542de-6bfd-4094-b8a6-3c6022f1eb21) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/30147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/24487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68484 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/100526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/24677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127885 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45554 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/34377 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/98681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45918 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/102596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/98465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/43913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/21912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18904 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45424 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51102 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44888 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48234 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46574 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/237d2cbee03f3628c325ccd6a4d289e2a447d325. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49864

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ac128b7fe3d0cd916eb5359537d4164d38f312b0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49864

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/12356a825f1d591a16934e169cd0801cfede1332. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49864

AriYoung00 commented on Source/WebCore/contentextensions/ContentExtensionActions.h:
```diff
@@ -233,18 +233,42 @@ struct WEBCORE_EXPORT RedirectAction {
+
+    static ReportIdentifierAction deserialize(std::span<const uint8_t> span)
+    {
+        return { ActionWithStringMetadata<ReportIdentifierAction>::deserialize(span.subspan(sizeof(identifier))).string, reinterpretCastSpanStartTo<double>(span) };
```
We could use the value returned by `ActionWithStringMetadata<ReportIdentifierAction>::deserialize` here instead of just pulling out the string:
```suggestion
        ReportIdentifierAction value = ActionWithStringMetadata<ReportIdentifierAction>::deserialize(span.subspan(sizeof(identifier)));
        value.identifier = reinterpretCastSpanStartTo<double>(span);
        return value;
```

elijahsawyers committed:
dNR: include the rest of the parameters on the MatchedRuleInfoDebug object
https://bugs.webkit.org/show_bug.cgi?id=297881
rdar://157880177

Reviewed by Brian Weinstein.

This patch makes the necessary changes to include the rest of the information on the MatchedRuleInfo
object that's passed to the onRuleMatchedDebug listeners.

Most of the request properties could easily be set with the information that is already available in
ContentExtensionsBackend::processContentRuleListsForLoad. However, we are missing the rulesetId
property of the rule object, and it's not readily available because compiled content rule lists do
not track which ruleset a rule belongs to. This patch updates the ReportIdentifierAction to inherit
from ActionWithStringMetadata so that we can serialize its ruleset identifier too in addition to the
rule identifier.

For cleaner code, we've chosen to serialize the ruleset identifier repeatedly because most popular
dNR web extensions appear to choose fairly short strings for the IDs. If this proves to consume too
much memory in the future, we can update the strings to be serialized once and store an index to it
in the ReportIdentifierAction.

The identifiers of static rulesets are set in the manifest. Session and dynamic rules have the
identifier, "_session" and "_dynamic," respectively.

For information about the properties that we've added, take a look at:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest/onRuleMatchedDebug#listener

It's worth noting that we aren't including the parentDocumentId right now because there is some work
to be done to support it with site isolation turned on. That work is tracked with: rdar://159289161

Added a new tests and updated existing tests to validate this fix.

* Source/WebCore/contentextensions/ContentExtensionActions.h:
(WebCore::ContentExtensions::ReportIdentifierAction::ReportIdentifierAction):
(WebCore::ContentExtensions::ReportIdentifierAction::isolatedCopy const):
(WebCore::ContentExtensions::ReportIdentifierAction::isolatedCopy):
(WebCore::ContentExtensions::ReportIdentifierAction::serialize const):
(WebCore::ContentExtensions::ReportIdentifierAction::deserialize):
(WebCore::ContentExtensions::ReportIdentifierAction::serializedLength):
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::loadRuleIdentifier):
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
* Source/WebCore/contentextensions/ContentRuleListMatchedRule.h:
(WebCore::ContentRuleListMatchedRule::ContentRuleListMatchedRule):
* Source/WebCore/loader/ResourceLoadInfo.cpp:
(WebCore::ContentExtensions::resourceTypeToStringForMatchedRule):
* Source/WebCore/loader/ResourceLoadInfo.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules):
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
(WebKit::WebExtensionContext::handleContentRuleListMatchedRule):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::handleContentRuleListMatchedRule):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
(-[_WKWebExtensionDeclarativeNetRequestRule _webKitRuleWithWebKitActionType:chromeActionType:condition:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm:
(WebKit::for):
(+[_WKWebExtensionDeclarativeNetRequestTranslator jsonObjectsFromData:errorStrings:]):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIDeclarativeNetRequest::updateDynamicRules):
(WebKit::WebExtensionAPIDeclarativeNetRequest::updateSessionRules):
* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::TEST_F(ContentExtensionTest, ReportIdentifierAction)):

Canonical link: https://commits.webkit.org/299333@main
